### PR TITLE
Add support for creating a GKE cloud credential

### PR DESCRIPTION
**Pull request by cmurphy**
This patch adds the ability to create and store a CloudCredential via
the /v3/cloudcredentials API that can be used to drive GKE clusters.

By adding it to the machine driver controller, Rancher can expose this
credential type as a dynamic schema in the same way that other cloud
credentials are supported. However, since GKE is not a real machine
driver, we have to fake it by creating dummy flags for a pretend machine
driver binary.

The credential is identified by the private field "authEncodedJson"
where the value is an *escaped* string representing a JSON object for
the service account data obtained from Google. If the credential object
is provided as a regular JSON object rather than an escaped string,
Rancher will end up converting into a Go map and then to a string
representation of the map before storing it in a Secret, rather than
directly storing the JSON object that is needed to authenticate with
Google. The result is that creating a cloud credential of this type will
look like this:

```
HTTP/1.1 POST /v3/cloudcredentials

{
  "googlecredentialConfig": {
    "authEncodedJson": "{\"type\": \"service_account\",
      \"project_id\": \"my-project\",
      \"private_key_id\": \"abcdef\",
      \"private_key\": \"-----BEGIN PRIVATE KEY-----\\nABCdef==\\n-----END PRIVATE KEY-----\\n\",
      \"client_email\": \"user@my-project.iam.gserviceaccount.com\",
      \"client_id\": \"123456\",
      \"auth_uri\": \"https://accounts.google.com/o/oauth2/auth\",
      \"token_uri\": \"https://oauth2.googleapis.com/token\",
      \"auth_provider_x509_cert_url\": \"https://www.googleapis.com/oauth2/v1/certs\",
      \"client_x509_cert_url\": \"https://www.googleapis.com/robot/v1/metadata/x509/user%40project.iam.gserviceaccount.com\"}"
  }
}
```

#30981

### CODE REVIEW

File: pkg/data/management/machinedriver_data.go
Comment (thedadams) [reply pkg/data/management/machinedriver_data.go:31]: The name of the driver should be `google` in case we add more general support in the future.

File: pkg/controllers/management/drivers/nodedriver/machine_driver.go
Comment (thedadams) [reply pkg/controllers/management/drivers/nodedriver/machine_driver.go:190]: I believe this is the only place the driver is called here. This is just to get the flags for creation.

The name of the flag should be `google-auth-encoded-json` to make it compatible with machine (in case we add more general support).

Comment (thedadams) [reply pkg/controllers/management/drivers/nodedriver/machine_driver.go:190]: If the changes do get merged into `rancher-machine`, then this block can be removed.

**Review by thedadams — CHANGES_REQUESTED** [review]

File: pkg/controllers/management/drivers/nodedriver/machine_driver.go
Comment (cmurphy) [reply pkg/controllers/management/drivers/nodedriver/machine_driver.go:190]: Where does the string `google-auth-encoded-json` come from? I used `credential` because that's currently how kontainer-engine identifies it

https://github.com/rancher/kontainer-engine/blob/89626b028c6a0cb1eeb399c9e7dc5404286efbb4/drivers/gke/gke_driver.go#L323 

https://github.com/rancher/kontainer-engine/blob/89626b028c6a0cb1eeb399c9e7dc5404286efbb4/drivers/gke/gke_driver.go#L846

File: pkg/data/management/machinedriver_data.go
Comment (cbron) [reply pkg/data/management/machinedriver_data.go:31]: `google` or `gcp` ? I'd vote second one.

Comment (cmurphy) [reply pkg/data/management/machinedriver_data.go:31]: +1 for `gcp`

Comment (thedadams) [reply pkg/data/management/machinedriver_data.go:31]: It's not a choice, unfortunately. This code uses rancher-machine, so if we are going to support GCP in the future then we have to use the driver's name in rancher-machine which is `google`.

Comment (cmurphy) [reply pkg/data/management/machinedriver_data.go:31]: It's not using rancher-machine though, it's purposely working around using rancher-machine. Don't we want to be able to decouple cloud credentials from machine provisioning?

Comment (cbron) [reply pkg/data/management/machinedriver_data.go:31]: We do, but there's a decent chance we'll add GCP VM's to rancher-machine in near future.

**Comment by cmurphy** [comment]
Fixed the driver name and the flag name. Also note the changed commit message and PR description: the JSON token credential needs to be provided as an escaped string, not a direct JSON object, otherwise the GKE operator will have a hard time extracting the object from the Secret.

**Review by thedadams — DISMISSED** [review]

**Review by cbron — COMMENTED** [review]
lgtm, but waiting on review from Kinara

File: pkg/controllers/management/node/controller.go
Comment (kinarashah) [inline pkg/controllers/management/node/controller.go:70]: note: this won't do anything at the moment. @dramich mentioned long term goal is to add a driver so can keep it as is

File: pkg/data/management/machinedriver_data.go
Comment (kinarashah) [inline pkg/data/management/machinedriver_data.go:99]: we don't need to whitelist `local://` since we're not going call rancher for google driver (thanks @dramich)

Comment (dramich) [reply pkg/data/management/machinedriver_data.go:99]: @cmurphy To expand on this the whitelist is URLs that rancher will proxy for the UI. The UI calls to providers through rancher to talk to the provider API. Depending on the provider it could be getting instance sizes, zones, etc. For Google, there will not be a UI portion that needs to talk to google so we don't want to whitelist a URL that rancher will not be using.

File: pkg/controllers/management/drivers/nodedriver/machine_driver.go
Comment (dramich) [inline pkg/controllers/management/drivers/nodedriver/machine_driver.go:195]: Can we move this inside the `getCreateFlagsForDriver` call? We still want the flags for the "driver" so shortcut in there instead of here.

File: pkg/data/management/machinedriver_data.go
Comment (cbron) [reply pkg/data/management/machinedriver_data.go:99]: Do you know if this whitelist is only for machine ? I believe the UI will need to make calls to gcp to get data for GKE, but that might be a different mechanism.

Comment (dramich) [reply pkg/data/management/machinedriver_data.go:99]: Same whitelist technically. If the UI needs to do that then something needs to be whitelisted, but it should probably be `container.googleapis.com` based on https://cloud.google.com/kubernetes-engine/docs/reference/rest

Comment (vincent99) [reply pkg/data/management/machinedriver_data.go:99]: Machine is running arbitrary code and can call whatever it wants; the whitelist is only for the /meta/proxy the UI uses

File: pkg/data/management/machinedriver_data.go
Comment (cbron) [reply pkg/data/management/machinedriver_data.go]: Follow up: yes UI will handle validation (both that field is required and creds are valid), but through meta proxy api calls.

File: pkg/data/management/machinedriver_data.go
Comment (cbron) [reply pkg/data/management/machinedriver_data.go:99]: We are aiming at using the GKE specific meta proxy for now which does not need this whitelist, only the generic `meta/proxy` does. We can leave this for now as it shouldn't hurt anything and we may support GCP VMs in future.

Comment (dramich) [reply pkg/data/management/machinedriver_data.go:99]: If we don't need it we should delete it. It's also not the correct URL for GCE.

Comment (cbron) [reply pkg/data/management/machinedriver_data.go:99]: That works also. Is that not right URL though ? Minus the `www` part:  `compute.googleapis.com`

File: pkg/controllers/management/drivers/nodedriver/machine_driver.go
Comment (cmurphy) [reply pkg/controllers/management/drivers/nodedriver/machine_driver.go]: done

File: pkg/data/management/machinedriver_data.go
Comment (cmurphy) [reply pkg/data/management/machinedriver_data.go]: It depends on which API is being used, I had been thinking this whitelist would be used to proxy auth requests and I had thought that `www.googleapis.com` would be used for that, but it might actually be `accounts.google.com` or `oauth2.googleapis.com`, I'm not sure. Removing it for now.

File: pkg/data/management/machinedriver_data.go
Comment (dramich) [inline pkg/data/management/machinedriver_data.go:99]:
```suggestion
	if err := addMachineDriver(GoogleDriver, "local://", "", "", nil, true, true, management); err != nil {
```

Comment (cmurphy) [reply pkg/data/management/machinedriver_data.go]: fixed

**Review by dramich — APPROVED** [review]

**Review by cbron — APPROVED** [review]
Lgtm. @kinarashah we are going to merge now to un-block some other work, lmk if you see anything in here you'd like addressed.

**Pull request merged**
